File: src/main/java/jug/java/Company.java
package jug.java;

import javax.annotation.Nonnull;
import java.util.Objects;
import java.util.stream.IntStream;
import java.util.stream.Stream;

public final class Company {

    @Nonnull
    private final String name;

    @Nonnull
    private static IntStream getCodes() {
        return IntStream.range(1_000, 50_001);
    }

    public Company(@Nonnull final String name) {
        Objects.requireNonNull(name);
        if (name.isEmpty()) {
            throw new IllegalArgumentException(("Empty name"));
        }
        this.name = name;
    }

    @Nonnull
    public String getName() {
        return name;
    }

    @Nonnull
    public Stream<Train> getTrains() {
        return getCodes().mapToObj((code) -> Util.makeTrain(this, code));
    }
}

File: src/main/kotlin/jug/kotlin/Company.kt
package jug.kotlin

class Company(val name: String) {

    init {
        require(name.isNotEmpty()) { "Empty name" }
    }

    val trains: Sequence<Train>
        get() = codes.asSequence().map { code -> makeTrain(this, code) }

    private companion object {
        val codes = 1_000..50_000
    }
}
